### PR TITLE
New diagnostics support RZ

### PR DESCRIPTION
**Pull request by MaxThevenet**
Merge after #835.

This PR uses the new vector of functors approach to allow diagnostics in RZ, recovering the previous behavior. More specifically:
- For the fields requested by the user, functor `CellCenterFunctor` performs the mode averaging before cell-centering.
- Anyhow, all modes of all fields are cell-centered and dumped, with the same naming convention as before (`Er_1_imag` etc.).

This is done by
- Adding the names `Er_1_imag` etc. to `varnames`, that contains the list of names of output fields
- Modifying `CellCenterFunctor` so that it also performs mode-averaging in RZ
- Adding functors to dump all modes at the end of `all_field_functors`.

In the previous implementation, the number of elements in `all_field_functors` and in the vector of output names `varnames` were equal to the number of components in the output MultiFab `mf_avg`. This PR changes this behavior as ` element in `all_field_functors` can now write into multiple components of `mf_avg`. This was done so because all RZ modes are components of MultiFabs `Efield_aux` etc. As a result
```
all_field_functors.size() <= varnames.size()
varnames.size() = mf_avg.nComp()
```
There are a few lines in the code to handle (and check) this indexing.

### CODE REVIEW

**Comment by dpgrote** [comment]
Overall this looks good - thanks for the PR!

I find the use of ncomp_src and ncomp_dst to be confusing. I think the code would be more clear if  these were not used but instead use array sizes directly. Is there are reason that AddRZModesToOutputNames is called before the loop at line 66, and not after, by the call to AddRZModesToInputsFunctors (or incorporated into AddRZModesToInputsFunctors)? If you did this, then the loop at line 70 could be `for (int comp=0; comp<varnames.size(); comp++){` and you wouldn't need the variable ncomp_except_cylindrical_modes.

Also, this is always writing out all of the RZ modes, independent of what is in fields_to_plot. This is probably fine, just different than what it is now.

**Comment by MaxThevenet** [comment]
Thanks for your review, these are good points, I implemented them in the last commit.

It is true that we are outputting all modes of E B and J, i.e., potentially more than before, but I would like to keep it that way, we still have enough data to check correctness.

**Comment by dpgrote** [comment]
This looks much better - thanks!

I have more comments.
in CellCenterFunctor, mode_avg should be mode_sum since it is a sum of modes. Change comments as well.
I think it would be cleaner to set up the RZ varnames in the same loops setting up the all_field_functors. This would ensure that the two arrays have the same ordering.

**Comment by MaxThevenet** [comment]
> I think it would be cleaner to set up the RZ varnames in the same loops setting up the all_field_functors

Do you mean merging functions `AddRZModesToOutputNames` and `AddRZModesToInputsFunctors`?

**Comment by dpgrote** [comment]
Yes, the two functions could be merged. Both have the same sets of loops and are adding field functors and var names for the same diagnostics.

The new variable name convertRZmodes2cartesian is good.

**Comment by MaxThevenet** [comment]
ok, done.

**Comment by dpgrote** [comment]
Thanks! Looks good, though there's extraneous indentation in the calls to CellCenterFunctor, and the warpx instance is no longer needed in AddRZModesToOutputNames

**Comment by MaxThevenet** [comment]
Thanks, done, just updated the PR. I don't see any indentation problem, and emacs agrees with me. If you still see one, could you put comments in the code directly so that I know where the issue is?

**Comment by dpgrote** [comment]
Oops, never mind about the indentations. I was misreading the code.

**Comment by MaxThevenet** [comment]
Could we then merge the PR when tests pass?

File: Source/Diagnostics/CellCenterFunctor.H [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+{
+public:
+    /** Constructor.
+     * \param[in] mf_src source multifab.
+     * \param[in] lev level of multifab. Used for averaging in rz.
+     * \param[in] convert_RZmodes2cartesian (in cylindrical) whether to
+     *            sum all modes in mf_src before cell-centering into dst multifab.
+     * \param[in] ncomp Number of component of mf_src to cell-center in dst multifab.
+     */
+    CellCenterFunctor(amrex::MultiFab* mf_src, int lev,
```
Comment (atmyers) [inline Source/Diagnostics/CellCenterFunctor.H]: can this pointer be const?

File: Source/Diagnostics/CellCenterFunctor.H [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+     * of a MultiFab and writes into one single component.
+     *
+     * \param[out] mf_dst output MultiFab where the result is written
+     * \param[in] dcomp first component of mf_dst in which cell-centered
+     *            data is stored
+     */
+    virtual void operator()(amrex::MultiFab& mf_dst, int dcomp) const override final;
+private:
+    /** pointer to source multifab (can be multi-component) */
+    amrex::MultiFab* m_mf_src;
```
Comment (atmyers) [inline Source/Diagnostics/CellCenterFunctor.H]: likewise here

File: Source/Diagnostics/CellCenterFunctor.cpp [hunk 3]
```diff
@@ -0,0 +1,38 @@
+#include "CellCenterFunctor.H"
+#include "Average.H"
+
+using namespace amrex;
+
+CellCenterFunctor::CellCenterFunctor(amrex::MultiFab* mf_src, int lev,
```
Comment (atmyers) [inline Source/Diagnostics/CellCenterFunctor.cpp]: and here?

File: Source/Diagnostics/CellCenterFunctor.H (hunk 1 above)
Comment (MaxThevenet) [reply Source/Diagnostics/CellCenterFunctor.H]: Yes, thanks, done

File: Source/Diagnostics/CellCenterFunctor.H (hunk 2 above)
Comment (MaxThevenet) [reply Source/Diagnostics/CellCenterFunctor.H]: Yes, thanks, done

File: Source/Diagnostics/CellCenterFunctor.cpp (hunk 3 above)
Comment (MaxThevenet) [reply Source/Diagnostics/CellCenterFunctor.cpp]: Yes, thanks, done

**Pull request merged**